src/query/functions/temporal/rate.go

```diff
@@ -36,36 +36,167 @@ const (
 	// IDeltaType calculates the difference between the last two values in the time series.
 	// IDeltaTemporalType should only be used with gauges.
 	IDeltaType = "idelta"
-)
 
-// NewRateOp creates a new base temporal transform for rate functions
-func NewRateOp(args []interface{}, optype string) (transform.Params, error) {
-	if optype == IRateType || optype == IDeltaType {
-		return newBaseOp(args, optype, newRateNode, nil)
-	}
+	// RateType calculates the per-second average rate of increase of the time series.
+	RateType = "rate"
 
-	return nil, fmt.Errorf("unknown rate type: %s", optype)
+	// DeltaType calculates the difference between the first and last value of each time series.
+	DeltaType = "delta"
+
+	// IncreaseType calculates the increase in the time series.
+	IncreaseType = "increase"
+)
+
+type rateProcessor struct {
+	isRate, isCounter bool
+	rateFn            rateFn
 }
 
-func newRateNode(op baseOp, controller *transform.Controller, opts transform.Options) Processor {
-	isRate := op.operatorType == IRateType
+func (r rateProcessor) Init(op baseOp, controller *transform.Controller, opts transform.Options) Processor {
+	var (
+		isRate, isCounter bool
+		rateFn            = standardRateFunc
+	)
+
+	switch op.operatorType {
```

Comment:
move this logic into NewRateOp. Init is called during actual execution, we need to make these decisions at the time of generating the plan.